### ISSUE TITLE
Some doubts about the nerf training forward and backward

### ISSUE THREAD

**Issue by lzhnb** [NVlabs/instant-ngp#614]
For each training step, you allocate the `max_sample` memory for the `coords`, which stores the positions of the sample points on the batch rays. And the `max_sample` equals to **4096 * 1024** means that **1024 sample points for 4096 batch rays**.
https://github.com/NVlabs/instant-ngp/blob/b97faed27aeacb6bfd91c6e7171d7e952e917e67/src/testbed_nerf.cu#L2861-L2899 

After allocating memory, you get the sample points with the help of `density_grid` in [generate_training_samples_nerf](https://github.com/NVlabs/instant-ngp/blob/b97faed27aeacb6bfd91c6e7171d7e952e917e67/src/testbed_nerf.cu#L1058). In the sampling, the sample is sparse, which you assume the sparse ratio is around `1/16`, and set the `target_batch_size` to be 1/16 of `max_sample`.
In `generate_training_samples_nerf`, you query the `density_grid` and skip the empty sample points. You only store the positions of the valid sample points in `coords_out` and padding for the empty sample points.
In my opinion, you realize the **compaction of samples in dense buffers for efficient execution** in `compute_cam_gradient_train_nerf` as you said in the **appendix C**. Right? But this part really confuses me. You first [inference](https://github.com/NVlabs/instant-ngp/blob/b97faed27aeacb6bfd91c6e7171d7e952e917e67/src/testbed_nerf.cu#L2973) using the `coords_matrix` as input to predict the rgbsigma for valid sample points and padding. Then, you conduct [compute_loss_kernel_train_nerf](https://github.com/NVlabs/instant-ngp/blob/b97faed27aeacb6bfd91c6e7171d7e952e917e67/src/testbed_nerf.cu#L2979) to compute loss as `dloss_dmlp_out` and compacted the valid sample points. Finally, you conduct the [forward](https://github.com/NVlabs/instant-ngp/blob/b97faed27aeacb6bfd91c6e7171d7e952e917e67/src/testbed_nerf.cu#L3044) and [backward](https://github.com/NVlabs/instant-ngp/blob/b97faed27aeacb6bfd91c6e7171d7e952e917e67/src/testbed_nerf.cu#L3045). The forward use the compacted coords input and the backward use the compacted gradient `dloss_dmlp_out`.
- In summary, you conduct **1 inference, 1 forward and 1 backward** for each nerf training step, right? You use all valid sample (1024 samples points per ray) points for **inference** to ensure the correct predicted rgb and the correct loss difference (if the valid sample points **more than 4096 / 16**), right?  For `forward` and `backward`, you only focus on the **compacted samples (64 samples for 4096 rays)** to realize optimization, right?

https://github.com/NVlabs/instant-ngp/blob/b97faed27aeacb6bfd91c6e7171d7e952e917e67/src/testbed_nerf.cu#L1174-L1217

**Comment by MarvinChung** [comment]
Hello @lzhnb , I have also read the code and these are my opinions. 

I don't think that ``max_sample`` means that 1024 sample points for 4096 batch rays.  
In my opinion,``max_sample`` preallocates the memory for all coordinates.  The number of coordinates are unknown since it depends on the density grid.  Therefore, a ray may have more than 1024 sample points or even zero points.  

This means than the numbers of coordinates may even exceed the ``max_sample``. 
The code then clip those points in the ``generate_training_samples_nerf`` function.  
https://github.com/NVlabs/instant-ngp/blob/899e1410fe69654a7d2c82398c9644c8ceac9cc8/src/testbed_nerf.cu#L1188-L1192

You refer compaction in the appendix C. I think that you have read the draft version of the paper. The latest paper refer compaction in the appendix E. You can find it on the main page of this repo. 

I agree with you that the code conducts 1 inference, 1 forward and 1 backward for each nerf training step.
Also the ``forward`` and ``backward``  only focus on the compacted samples to realize optimization.

I think this question will be more suitable to put in [Discussion](https://github.com/NVlabs/instant-ngp/discussions) for others to reply their opinions before the reply from the author.

**Comment by MarvinChung** [comment]
After reading the code, I have my own question.

Q: How much speed gain when compacting the coordinates?

In order to compact the coordinates, it needs additional GPU memory to store the compacted coordinates and requires ``atomicAdd`` in the ``compute_loss_kernel_train_nerf`` function which causes additional overhead. 
The benefit of compacting the coordinates is that less coordinates and only valid ones are passed to the ``forward`` and ``backward``. 
I am curious about the speed gain from compacting the coordinates.

**Comment by lzhnb** [comment]
> Hello @lzhnb , I have also read the code and these are my opinions.
> 
> I don't think that `max_sample` means that 1024 sample points for 4096 batch rays. In my opinion,`max_sample` preallocates the memory for all coordinates. The number of coordinates are unknown since it depends on the density grid. Therefore, a ray may have more than 1024 sample points or even zero points.
> 
> This means than the numbers of coordinates may even exceed the `max_sample`. The code then clip those points in the `generate_training_samples_nerf` function.
> 
> https://github.com/NVlabs/instant-ngp/blob/899e1410fe69654a7d2c82398c9644c8ceac9cc8/src/testbed_nerf.cu#L1188-L1192
> 
> You refer compaction in the appendix C. I think that you have read the draft version of the paper. The latest paper refer compaction in the appendix E. You can find it on the main page of this repo.
> 
> I agree with you that the code conducts 1 inference, 1 forward and 1 backward for each nerf training step. Also the `forward` and `backward` only focus on the compacted samples to realize optimization.
> 
> I think this question will be more suitable to put in [Discussion](https://github.com/NVlabs/instant-ngp/discussions) for others to reply their opinions before the reply from the author.

Thanks for your valuable insights👍  :)

**Comment by lzhnb** [comment]
> After reading the code, I have my own question.
> 
> Q: How much speed gain when compacting the coordinates?
> 
> In order to compact the coordinates, it needs additional GPU memory to store the compacted coordinates and requires `atomicAdd` in the `compute_loss_kernel_train_nerf` function which causes additional overhead. The benefit of compacting the coordinates is that less coordinates and only valid ones are passed to the `forward` and `backward`. I am curious about the speed gain from compacting the coordinates.

I have similar confusion. How can it be so fast and memory-efficient?(CUDA is all you need)😂

Maybe the forward and backward(building a computing graph in tiny-cuda-nn engine?) are the most time-consuming. So, compacting the coordinates may realize impressive acceleration🤔?
